packages/grafana-runtime/src/utils/DataSourceWithBackend.ts

```diff
@@ -0,0 +1,88 @@
+import {
+  DataSourceApi,
+  DataQueryRequest,
+  DataQueryResponse,
+  DataSourceInstanceSettings,
+  DataQuery,
+  DataSourceJsonData,
+} from '@grafana/data';
+import { Observable, from } from 'rxjs';
+import { config } from '..';
+import { getBackendSrv } from '../services';
+
+// Ideally internal (exported for consistency)
+const ExpressionDatasourceID = '__expr__';
+
+export class DataSourceWithBackend<
+  TQuery extends DataQuery = DataQuery,
+  TOptions extends DataSourceJsonData = DataSourceJsonData
+> extends DataSourceApi<TQuery, TOptions> {
+  constructor(instanceSettings: DataSourceInstanceSettings<TOptions>) {
+    super(instanceSettings);
+  }
+
+  /**
+   * Ideally final -- any other implementation would be wrong!
+   */
+  query(request: DataQueryRequest): Observable<DataQueryResponse> {
+    const { targets, intervalMs, maxDataPoints } = request;
+    let range = request.range;
+
+    let expressionCount = 0;
+    const orgId = config.bootData.orgId;
+    const queries = targets.map(q => {
+      if (q.datasource === ExpressionDatasourceID) {
+        expressionCount++;
+        return {
+          ...q,
+          datasourceId: this.id,
+          orgId,
+        };
+      }
+      const dsName = q.datasource && q.datasource !== 'default' ? q.datasource : config.defaultDatasource;
+      const ds = config.datasources[dsName];
+      if (!ds) {
+        throw new Error('Unknown Datasource: ' + q.datasource);
+      }
+      return {
+        ...q,
+        datasourceId: ds.id,
+        intervalMs,
+        maxDataPoints,
+        orgId,
+      };
+    });
+
+    const body: any = {
+      expressionCount,
+      queries,
+    };
+    if (range) {
+      body.range = range;
+      body.from = range.from.valueOf().toString();
+      body.to = range.to.valueOf().toString();
+    }
+
+    const req: Promise<DataQueryResponse> = getBackendSrv()
+      .post('/api/ds/query', body)
+      .then((rsp: any) => {
+        return this.toDataQueryResponse(rsp);
+      });
+    return from(req);
+  }
+
+  /**
+   * This makes the arrow libary loading async.
+   */
+  async toDataQueryResponse(rsp: any): Promise<DataQueryResponse> {
+    const { resultsToDataFrames } = await import(
+      /* webpackChunkName: "apache-arrow-util" */ '@grafana/data/src/dataframe/ArrowDataFrame'
+    );
+    return { data: resultsToDataFrames(rsp) };
```

Comment:
why would this not work for external plugins?

Reply:
oh, I see the import from src dir. Think that will work after @dprokop merges the package namespace PR (it will make importing sub paths without src dir work from app and other places)

Reply:
Yes, as @torkelo said, when the namespacing work is ready this should be work as expected. I'm gonna aim for making the @grafana/data namespacing priority

Reply:
I think this should work for external plugins as well, if an external plugin extend DataSourceWithBackend they will get the version we have built (using webpack), so if it works for internal plugins it should work for external ones

Reply:
Ref https://github.com/grafana/grafana/pull/21326#issuecomment-573075584 @torkelo I think your idea with exposing arrow via grafana-runtime service is not a bad one. The drawback is thought we won't take much advantage from async loading

Reply:
@ryantxu I think this PR should fix your issue with apache-arrow and PhantomJS: https://github.com/grafana/grafana/pull/21677